### PR TITLE
Snapshot: add new auto case for snapshot-list

### PR DESCRIPTION
**Pull request by meinaLi**
Automate case VIRT-16603 - [Disk snapshot][External] Get guest snapshot list with option

### CODE REVIEW

**Comment by meinaLi** [comment]
```
# avocado run --vt-type libvirt --test-runner=runner --vt-machine-type q35 snapshot_list.with_options
JOB ID     : c1cb5ada2b7bc3bd1b02a88efc4236ff5d269003
JOB LOG    : /root/avocado/job-results/job-2024-08-08T04.38-c1cb5ad/job.log
 (1/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.roots: PASS (34.97 s)
 (2/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.leaves: PASS (28.24 s)
 (3/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.no_leaves.with_external: PASS (34.83 s)
 (4/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.no_leaves.with_inactive: PASS (29.13 s)
 (5/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.no_leaves.with_active: PASS (27.08 s)
 (6/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.no_leaves.with_disk_only: PASS (33.51 s)
 (7/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.from_and_descendants: PASS (29.30 s)
 (8/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.topological: PASS (27.31 s)
 (9/9) type_specific.io-github-autotest-libvirt.snapshot_list.with_options.parent: PASS (36.40 s)
RESULTS    : PASS 9 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2024-08-08T04.38-c1cb5ad/results.html
JOB TIME   : 282.82 s
```

**Comment by meinaLi** [comment]
@nanli1 Can you help to review this PR? Thanks.

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    list_options = params.get("list_options")
+    expected_list = eval(params.get("expected_list", "[]"))
+    expected_dict = eval(params.get("expected_dict", "{}"))
+    special_check = "yes" == params.get("special_check", "no")
+    vm = env.get_vm(vm_name)
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    bkxml = vmxml.copy()
+    virsh_dargs = {"debug": True, "ignore_status": False}
+    test_obj = snapshot_base.SnapshotTest(vm, test, params)
+    libvirt_version.is_libvirt_feature_supported(params)
```
Comment (nanli1) [inline libvirt/tests/src/snapshot/snapshot_list_with_options.py:80]: Version checking could me more forward ,maybe L68

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        Create snapshots for guest.
+
+        :params vm_name: the vm name
+        :params snap_names: the created snapshot names
+        """
+        test.log.info("TEST_SETUP: Prepare shutoff guest and create a snapshot.")
+        virsh.snapshot_create_as(vm_name, "%s --disk-only" % snap_names[0], **virsh_dargs)
+        test.log.info("TEST_SETUP: Prepare running guest and create a disk only snapshot.")
+        if not vm.is_alive():
+            vm.start()
```
Comment (nanli1) [inline libvirt/tests/src/snapshot/snapshot_list_with_options.py:31]: We could login vm to avoid some unpredictable issues

**Review by nanli1 — APPROVED** [review]

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py (hunk 1 above)
Comment (meinaLi) [reply libvirt/tests/src/snapshot/snapshot_list_with_options.py]: I think moving to L76 is more better after variant definition. So I've updated it to there.

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py (hunk 2 above)
Comment (meinaLi) [reply libvirt/tests/src/snapshot/snapshot_list_with_options.py:31]: Good point. I think it's a useful checkpoint. Finish to update now.

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    def check_snap_list(snap_list):
+        """
+        Check the list snapshots of guest to match expected ones.
+
+        :params snap_list: the snapshots created in setup
+        """
+        if special_check:
+            list_dict = {}
+            cmd = "virsh snapshot-list %s %s" % (vm_name, list_options)
+            list_result = process.run(cmd, shell=True, ignore_status=False).stdout_text.split()
```
Comment (Yingshun) [inline libvirt/tests/src/snapshot/snapshot_list_with_options.py:51]: For virsh commands, I think it's better to reuse the functions in virsh module. Can you plz try virsh.command or virsh.snapshot_list?

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            list_dict[list_result[0]] = ''
+            for num in range(1, len(list_result), 2):
+                key = list_result[num]
+                value = list_result[num+1]
+                list_dict[key] = value
+            for key, value in expected_dict.items():
+                if list_dict.get(key) != value:
+                    test.fail("Get incorrect result %s, the expected result is %s" %
+                              (list_result, expected_dict))
+            test.log.debug("Get expected result for snapshot-list options %s" %
```
Comment (Yingshun) [inline libvirt/tests/src/snapshot/snapshot_list_with_options.py:61]:
```suggestion
            test.log.debug("Get expected result for snapshot-list options %s",
```

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                           list_options)
+        for snap_name in expected_list:
+            if snap_name not in snap_list:
+                test.fail("Snapshot %s doesn't exist" % snap_name)
+            else:
+                test.log.info("Snapshot %s is in expected snapshot list." % snap_name)
```
Comment (Yingshun) [inline libvirt/tests/src/snapshot/snapshot_list_with_options.py:67]: '%' should be ','.

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py (hunk 3 above)
Comment (meinaLi) [reply libvirt/tests/src/snapshot/snapshot_list_with_options.py:51]: Actually I've tried to use virsh.snapshot_list before. But it's hard to deal with the result to match the expected list. Maybe you can have more better solution to deal with the cmd result?

Comment (Yingshun) [reply libvirt/tests/src/snapshot/snapshot_list_with_options.py:51]: Ok, can you plz tell me what the problem is? and can't 'virsh.command' be used?

Comment (meinaLi) [reply libvirt/tests/src/snapshot/snapshot_list_with_options.py]: Thanks to your detailed introduction. I changed it from process.run to virsh.command and the cases can pass.

File: libvirt/tests/src/snapshot/snapshot_list_with_options.py (hunk 5 above)
Comment (meinaLi) [reply libvirt/tests/src/snapshot/snapshot_list_with_options.py]: Updated.

**Review by Yingshun — APPROVED** [review]

**Pull request merged**